File: Step03_Arrays/Lec1_Easy/Program_09_brute.cpp
#include <bits/stdc++.h>
using namespace std;

// Find the union of two sorted arrays

int main() {
    int n1, n2;
    cin >> n1;
    vector<int> arr1(n1);
    for (int i = 0; i < n1; i++) {
        cin >> arr1[i];
    }
    cin >> n2;
    vector<int> arr2(n2);
    for (int i = 0; i < n2; i++) {
        cin >> arr2[i];
    }

    set<int> st;                // extra space taken
    for (auto element : arr1) { // O(n1 log n); n = size of set
        st.insert(element);
    }
    for (auto element : arr2) { // O(n2 log n); n = size of set
        st.insert(element);
    }

    vector<int> _union_;      // extra space taken
    for (auto element : st) { // O(n1 + n2)
        _union_.push_back(element);
    }

    for (auto element : _union_) {
        cout << element << " ";
    }
}

// Time-complexity  : O(n1 log n) + O(n2 log n) + O(n1+n2)
// Space-complexity : O(n1 + n2) + O(n1 + n2)

File: Step03_Arrays/Lec1_Easy/Program_09_brute.java
package Step03_Arrays.Lec1_Easy;

import java.util.*;

// Find the union of two sorted arrays

public class Program_09_brute {
    public static int[] getUnion(int[] arr1, int[] arr2) {

        // extra space taken
        Set<Integer> st = new TreeSet<>();

        // O(n1 log n); n = size of set
        for (int element : arr1) {
            st.add(element);
        }

        // O(n2 log n); n = size of set
        for (int element : arr2) {
            st.add(element);
        }

        // extra space taken
        int[] union = new int[st.size()];
        int i = 0;

        // O(n1 + n2)
        for (Integer element : st) {
            union[i++] = element;
        }

        return union;
    }

    public static void main(String[] args) {
        int[] arr1 = {1, 3, 3, 7, 9, 9, 9};
        int[] arr2 = {2, 4, 4, 4, 6, 8, 8, 8, 10};

        for (int element : getUnion(arr1, arr2)) {
            System.out.print(element + " ");
        }
    }
}

// Time-complexity  : O(n1 log n) + O(n2 log n) + O(n1+n2)
// Space-complexity : O(n1 + n2) + O(n1 + n2)
